src/float/add.rs

```diff
@@ -0,0 +1,268 @@
+                        - (implicit_bit << 3).0.leading_zeros()) as <$ty as Float>::Int;
+                    a_significand <<= shift as usize;
+                    a_exponent -= Wrapping(shift);
+                }
+            } else /* addition */ {
+                a_significand += b_significand;
+
+                // If the addition carried up, we need to right-shift the result and
+                // adjust the exponent:
+                if (a_significand & implicit_bit << 4).0 != 0 {
+                    let sticky = ((a_significand & one).0 != 0) as <$ty as Float>::Int;
+                    a_significand = a_significand >> 1 | Wrapping(sticky);
+                    a_exponent += one;
+                }
+            }
+
+            // If we have overflowed the type, return +/- infinity:
+            if a_exponent >= max_exponent { 
+                return (<$ty>::from_repr((inf_rep | result_sign).0));
+            }
+
+            if a_exponent.0 <= 0 {
+                // Result is denormal before rounding; the exponent is zero and we
+                // need to shift the significand.
+                let shift = one - a_exponent;
+                let sticky = ((a_significand << (bits - shift).0 as usize).0 != 0) as <$ty as Float>::Int;
+                a_significand = a_significand >> (shift.0 | sticky) as usize;
+                a_exponent = zero;
+            }
+
+            // Low three bits are round, guard, and sticky.
+            let round_guard_sticky = a_significand & Wrapping(0x7);
+
+            // Shift the significand into place, and mask off the implicit bit.
+            let mut result = a_significand >> 3 & significand_mask;
+
+            // Insert the exponent and sign.
+            result |= a_exponent << significand_bits.0 as usize;
+            result |= result_sign;
+
+            // Final rounding.  The result may overflow to infinity, but that is the
+            // correct result in that case.
+            if round_guard_sticky.0 > 0x4 { result += one; }
+            if round_guard_sticky.0 == 0x4 { result += result & one; }
+            return (<$ty>::from_repr(result.0));
+        }
+    }
+}
+
+add!(__addsf3: f32);
+add!(__adddf3: f64);
+
+// FIXME: Implement these using aliases
+#[cfg(target_arch = "arm")]
+#[cfg_attr(not(test), no_mangle)]
+pub extern fn __aeabi_dadd(a: f64, b: f64) -> f64 {
+    __adddf3(a, b)
+}
+
+#[cfg(target_arch = "arm")]
+#[cfg_attr(not(test), no_mangle)]
+pub extern fn __aeabi_dadd(a: f32, b: f32) -> f32 {
+    __addsf3(a, b)
+}
+
+#[cfg(test)]
+mod test {
+    use core::{f32, f64};
+    use qc::{U32, U64};
+    use float::Float;
+
+    // TODO: Add F32/F64 to qc so that they print the right values (at the very least)
+    quickcheck! {
+        fn addsf3(a: U32, b: U32) -> bool {
+            let (a, b) = (a.0, b.0);
+            let r = super::__addsf3(f32::from_repr(a), f32::from_repr(b));
+            r == f32::from_repr(a) + f32::from_repr(b)
```

Comment:
> I think we should leave strict equality for now and revisit this later if we find that it causes problems.

Sounds good to me.

> However I think we should be comparing the raw bits of the float, which means transmuting it to an u32 before comparing it

Yeah, I agree raw bits comparison is better. NaN has multiple, different representations, right? (as in there are a bunch of NaNs among the possible 2^32 single precision float values) Even raw bits equality may fail in the NaN case.